File: app/src/main/java/com/beautifulquran/domain/WordSearch.kt
package com.beautifulquran.domain

import com.beautifulquran.data.model.SurahWordSearchSection
import com.beautifulquran.data.model.WordSearchHit

/** Soft cap so a very common English gloss cannot flood the cover sheet. */
const val WORD_SEARCH_MAX_HITS = 400

/** Minimum trimmed query length before Quran-wide word search runs. */
const val WORD_SEARCH_MIN_QUERY_LENGTH = 2

/** How many ayah hits to show per surah before the expand line. */
const val WORD_SEARCH_PREVIEW_LIMIT = 3

/**
 * Lightweight word row used to build (and unit-test) Quran-wide search
 * without needing the Android SQLite wrapper.
 */
data class WordSearchIndexEntry(
    val surahId: Int,
    val ayahNumber: Int,
    val position: Int,
    val arabic: String,
    val arabicNorm: String,
    val translation: String,
    val translationLower: String,
    val transliteration: String,
    val transliterationLower: String,
    val ayahText: String,
    val ayahTranslation: String,
    val surahNameTransliteration: String,
    val surahNameArabic: String,
)

fun WordSearchIndexEntry.toHit(): WordSearchHit =
    WordSearchHit(
        surahId = surahId,
        ayahNumber = ayahNumber,
        position = position,
        arabic = arabic,
        translation = translation,
        transliteration = transliteration,
        ayahText = ayahText,
        ayahTranslation = ayahTranslation,
        surahNameTransliteration = surahNameTransliteration,
        surahNameArabic = surahNameArabic,
    )

/**
 * Returns true when [query] is long enough to run word search. Callers that
 * also treat `surah:ayah` as a jump reference should skip word search for
 * those queries themselves.
 */
fun isWordSearchQuery(query: String): Boolean {
    val trimmed = query.trim()
    return trimmed.length >= WORD_SEARCH_MIN_QUERY_LENGTH
}

/**
 * Scans [index] for Arabic (diacritic-insensitive), English gloss, or
 * transliteration substring matches. Results stay in Quranic order.
 *
 * When the SI ayah translation cannot show the matched English, the hit's
 * [WordSearchHit.ayahTranslation] is the same-ayah word-gloss line instead,
 * so the cover snippet always has the matched term in context.
 */
fun matchWordSearch(
    index: List<WordSearchIndexEntry>,
    query: String,
    maxHits: Int = WORD_SEARCH_MAX_HITS,
): List<WordSearchHit> {
    val trimmed = query.trim()
    if (!isWordSearchQuery(trimmed)) return emptyList()
    val arabicNorm = normalizeArabicForSearch(trimmed)
    val lower = trimmed.lowercase()
    val out = ArrayList<WordSearchHit>(minOf(64, maxHits))
    for (i in index.indices) {
        val entry = index[i]
        val hit = when {
            arabicNorm.length >= WORD_SEARCH_MIN_QUERY_LENGTH &&
                entry.arabicNorm.contains(arabicNorm) -> true
            entry.translationLower.contains(lower) -> true
            entry.transliterationLower.contains(lower) -> true
            else -> false
        }
        if (!hit) continue
        val base = entry.toHit()
        // Word glosses and SI ayah lines often disagree ("a resting place" vs
        // "a bed"). Prefer the gloss line only when it can host the highlight.
        val display = snippetDisplayText(entry, index, i, trimmed)
        out.add(if (display == entry.ayahTranslation) base else base.copy(ayahTranslation = display))
        if (out.size >= maxHits) break
    }
    return out
}

/**
 * SI ayah text when it can show the match; otherwise the same-ayah word-gloss
 * line when that can. Falls back to SI when neither hosts a highlight.
 */
internal fun snippetDisplayText(
    entry: WordSearchIndexEntry,
    index: List<WordSearchIndexEntry>,
    at: Int,
    query: String,
): String {
    if (highlightNeedle(entry.ayahTranslation, query, entry.translation) != null) {
        return entry.ayahTranslation
    }
    val glossLine = sameAyahGlossLine(index, at)
    if (highlightNeedle(glossLine, query, entry.translation) != null) return glossLine
    return entry.ayahTranslation
}

/** Space-joined English glosses for every word of the same ayah as [at]. */
internal fun sameAyahGlossLine(index: List<WordSearchIndexEntry>, at: Int): String {
    if (at !in index.indices) return ""
    val anchor = index[at]
    var lo = at
    while (
        lo > 0 &&
        index[lo - 1].surahId == anchor.surahId &&
        index[lo - 1].ayahNumber == anchor.ayahNumber
    ) {
        lo--
    }
    var hi = at
    while (
        hi + 1 < index.size &&
        index[hi + 1].surahId == anchor.surahId &&
        index[hi + 1].ayahNumber == anchor.ayahNumber
    ) {
        hi++
    }
    return (lo..hi).joinToString(" ") { index[it].translation }
}

/**
 * Groups flat hits into surah sections. Collapsed sections keep the first
 * [previewLimit] hits; expanded sections (ids in [expandedSurahIds]) keep all.
 */
fun sectionWordSearchHits(
    hits: List<WordSearchHit>,
    expandedSurahIds: Set<Int>,
    previewLimit: Int = WORD_SEARCH_PREVIEW_LIMIT,
): List<SurahWordSearchSection> {
    if (hits.isEmpty()) return emptyList()
    val grouped = linkedMapOf<Int, MutableList<WordSearchHit>>()
    for (hit in hits) {
        grouped.getOrPut(hit.surahId) { mutableListOf() }.add(hit)
    }
    return grouped.map { (surahId, surahHits) ->
        val expanded = surahId in expandedSurahIds
        val visible = if (expanded || surahHits.size <= previewLimit) {
            surahHits
        } else {
            surahHits.take(previewLimit)
        }
        val first = surahHits.first()
        SurahWordSearchSection(
            surahId = surahId,
            surahNameTransliteration = first.surahNameTransliteration,
            surahNameArabic = first.surahNameArabic,
            hits = visible,
            totalCount = surahHits.size,
            expanded = expanded,
        )
    }
}

/**
 * Builds spans for an ayah, highlighting the whitespace-token at 1-based
 * [position]. Falls back to highlighting every exact surface-form match of
 * [fallbackWord] when the token split does not line up.
 */
fun ayahHighlightSpans(
    ayahText: String,
    position: Int,
    fallbackWord: String,
): List<AyahTextSpan> {
    if (ayahText.isEmpty()) return emptyList()
    val tokens = ayahText.split(Regex("\\s+")).filter { it.isNotEmpty() }
    if (position in 1..tokens.size) {
        val spans = ArrayList<AyahTextSpan>(tokens.size * 2)
        tokens.forEachIndexed { index, token ->
            if (index > 0) spans.add(AyahTextSpan(" ", highlighted = false))
            spans.add(AyahTextSpan(token, highlighted = index + 1 == position))
        }
        return spans
    }
    if (fallbackWord.isEmpty()) {
        return listOf(AyahTextSpan(ayahText, highlighted = false))
    }
    val spans = ArrayList<AyahTextSpan>()
    var start = 0
    var i = ayahText.indexOf(fallbackWord)
    if (i < 0) return listOf(AyahTextSpan(ayahText, highlighted = false))
    while (i >= 0) {
        if (i > start) spans.add(AyahTextSpan(ayahText.substring(start, i), highlighted = false))
        spans.add(AyahTextSpan(fallbackWord, highlighted = true))
        start = i + fallbackWord.length
        i = ayahText.indexOf(fallbackWord, start)
    }
    if (start < ayahText.length) {
        spans.add(AyahTextSpan(ayahText.substring(start), highlighted = false))
    }
    return spans
}

/** How many whole words of context to keep on each side of a search hit. */
const val SNIPPET_WORDS_BEFORE = 8
const val SNIPPET_WORDS_AFTER = 14

/**
 * Builds spans for an English search snippet: a short window of text centered
 * on the match (query or word gloss), with the match highlighted. Used by the
 * cover-sheet word results so the gold term is always on-screen.
 */
fun englishTranslationHighlightSpans(
    ayahTranslation: String,
    query: String,
    wordGloss: String,
): List<AyahTextSpan> {
    if (ayahTranslation.isEmpty()) return emptyList()
    val needle = highlightNeedle(ayahTranslation, query.trim(), wordGloss.trim())
    val snippet = windowAroundMatch(
        text = ayahTranslation,
        needle = needle,
        wordsBefore = SNIPPET_WORDS_BEFORE,
        wordsAfter = SNIPPET_WORDS_AFTER,
    )
    if (needle == null) return listOf(AyahTextSpan(snippet, highlighted = false))
    return highlightAllOccurrences(snippet, needle)
}

/**
 * Trims [text] to roughly [wordsBefore]…[wordsAfter] words around the first
 * occurrence of [needle], adding an ellipsis when the ends were cut.
 * When [needle] is null or missing, returns [text] unchanged.
 */
internal fun windowAroundMatch(
    text: String,
    needle: String?,
    wordsBefore: Int = SNIPPET_WORDS_BEFORE,
    wordsAfter: Int = SNIPPET_WORDS_AFTER,
): String {
    if (needle.isNullOrEmpty() || text.isEmpty()) return text
    val matchStart = text.indexOf(needle, ignoreCase = true)
    if (matchStart < 0) return text
    val matchEnd = matchStart + needle.length
    val words = Regex("\\S+").findAll(text).toList()
    if (words.isEmpty()) return text
    val matchWord = words.indexOfFirst { mr ->
        matchStart < mr.range.last + 1 && matchEnd > mr.range.first
    }.let { if (it < 0) 0 else it }
    val from = (matchWord - wordsBefore).coerceAtLeast(0)
    val to = (matchWord + wordsAfter).coerceAtMost(words.lastIndex)
    val startChar = words[from].range.first
    val endChar = words[to].range.last + 1
    val core = text.substring(startChar, endChar).trim()
    val prefix = if (from > 0) "…" else ""
    val suffix = if (to < words.lastIndex) "…" else ""
    return prefix + core + suffix
}

/**
 * Prefers the typed query when it appears in the translation; otherwise the
 * word gloss. Returns null when neither can be found (case-insensitive).
 */
internal fun highlightNeedle(
    haystack: String,
    query: String,
    wordGloss: String,
): String? {
    if (query.isNotEmpty() && haystack.contains(query, ignoreCase = true)) {
        return query
    }
    if (wordGloss.isNotEmpty() && haystack.contains(wordGloss, ignoreCase = true)) {
        return wordGloss
    }
    // Gloss rows are often "(is) the Merciful" while the ayah says "the Merciful" —
    // try the longest whitespace token from the gloss that still appears.
    val tokens = wordGloss
        .split(Regex("[\\s,;:]+"))
        .map { it.trim().trim('(', ')', '[', ']', '"', '\'') }
        .filter { it.length >= 3 }
        .sortedByDescending { it.length }
    for (token in tokens) {
        if (haystack.contains(token, ignoreCase = true)) return token
    }
    return null
}

private fun highlightAllOccurrences(text: String, needle: String): List<AyahTextSpan> {
    val spans = ArrayList<AyahTextSpan>()
    var start = 0
    var i = text.indexOf(needle, ignoreCase = true)
    if (i < 0) return listOf(AyahTextSpan(text, highlighted = false))
    while (i >= 0) {
        if (i > start) spans.add(AyahTextSpan(text.substring(start, i), highlighted = false))
        val end = i + needle.length
        spans.add(AyahTextSpan(text.substring(i, end), highlighted = true))
        start = end
        i = text.indexOf(needle, start, ignoreCase = true)
    }
    if (start < text.length) {
        spans.add(AyahTextSpan(text.substring(start), highlighted = false))
    }
    return spans
}

data class AyahTextSpan(val text: String, val highlighted: Boolean)

File: web/src/domain/WordSearch.ts
/** Quran-wide word search — mirrors Android `domain/WordSearch.kt`. */

export const WORD_SEARCH_MAX_HITS = 400
export const WORD_SEARCH_MIN_QUERY_LENGTH = 2
export const WORD_SEARCH_PREVIEW_LIMIT = 3

export interface WordSearchHit {
  surahId: number
  ayahNumber: number
  position: number
  arabic: string
  translation: string
  transliteration: string
  ayahText: string
  ayahTranslation: string
  surahNameTransliteration: string
  surahNameArabic: string
}

export interface SurahWordSearchSection {
  surahId: number
  surahNameTransliteration: string
  surahNameArabic: string
  hits: WordSearchHit[]
  totalCount: number
  expanded: boolean
  hiddenCount: number
}

export interface WordSearchIndexEntry {
  surahId: number
  ayahNumber: number
  position: number
  arabic: string
  arabicNorm: string
  translation: string
  translationLower: string
  transliteration: string
  transliterationLower: string
  ayahText: string
  ayahTranslation: string
  surahNameTransliteration: string
  surahNameArabic: string
}

/**
 * Strips tashkeel / tatweel and unifies alef / ya variants so typed Arabic
 * can match Uthmani surface forms. Mirrors Android `normalizeArabicForSearch`
 * and `tools/build_db.py` `normalize_for_alignment`.
 *
 * Uses Arabic mark code-point ranges instead of `\p{M}` so index build over
 * ~77k words stays cheap on the main thread.
 */
export function normalizeArabicForSearch(input: string): string {
  if (!input) return input
  let out = ''
  for (let i = 0; i < input.length; i++) {
    let cp = input.charCodeAt(i)
    // Skip UTF-16 surrogate pairs' trail — Quran Arabic is BMP.
    if (cp >= 0xd800 && cp <= 0xdbff) {
      i++
      continue
    }
    if (cp === 0x0671 || cp === 0x0622 || cp === 0x0623 || cp === 0x0625) {
      cp = 0x0627
    } else if (cp === 0x0649) {
      cp = 0x064a
    } else if (cp === 0x0640) {
      continue
    }
    // Arabic diacritics / Quranic annotation marks (not full \p{M}).
    if (
      (cp >= 0x064b && cp <= 0x065f) ||
      cp === 0x0670 ||
      (cp >= 0x06d6 && cp <= 0x06ed) ||
      (cp >= 0x08d3 && cp <= 0x08ff)
    ) {
      continue
    }
    if (cp >= 0x0621 && cp <= 0x064a) out += String.fromCharCode(cp)
  }
  return out
}

export function isWordSearchQuery(query: string): boolean {
  return query.trim().length >= WORD_SEARCH_MIN_QUERY_LENGTH
}

export interface AyahReference {
  surah: number
  ayah: number | null
}

const ayahReferenceRegex = /^\s*(\d+)\s*:\s*(\d+)?\s*$/

export function parseAyahReference(query: string): AyahReference | null {
  const match = ayahReferenceRegex.exec(query)
  if (!match) return null
  const surah = Number(match[1])
  if (!Number.isFinite(surah)) return null
  const ayahText = match[2]
  if (ayahText == null || ayahText === '') return { surah, ayah: null }
  const ayah = Number(ayahText)
  if (!Number.isFinite(ayah)) return null
  return { surah, ayah }
}

/** Word search runs for typed queries that are not `surah:ayah` jumps. */
export function shouldRunWordSearch(query: string): boolean {
  if (!isWordSearchQuery(query)) return false
  return parseAyahReference(query.trim()) == null
}

export function toHit(entry: WordSearchIndexEntry): WordSearchHit {
  return {
    surahId: entry.surahId,
    ayahNumber: entry.ayahNumber,
    position: entry.position,
    arabic: entry.arabic,
    translation: entry.translation,
    transliteration: entry.transliteration,
    ayahText: entry.ayahText,
    ayahTranslation: entry.ayahTranslation,
    surahNameTransliteration: entry.surahNameTransliteration,
    surahNameArabic: entry.surahNameArabic,
  }
}

export function matchWordSearch(
  index: WordSearchIndexEntry[],
  query: string,
  maxHits = WORD_SEARCH_MAX_HITS,
): WordSearchHit[] {
  const trimmed = query.trim()
  if (!isWordSearchQuery(trimmed)) return []
  const arabicNorm = normalizeArabicForSearch(trimmed)
  const lower = trimmed.toLowerCase()
  const wantArabic = arabicNorm.length >= WORD_SEARCH_MIN_QUERY_LENGTH
  const out: WordSearchHit[] = []
  for (let i = 0; i < index.length; i++) {
    const entry = index[i]!
    const hit =
      (wantArabic && entry.arabicNorm.includes(arabicNorm)) ||
      entry.translationLower.includes(lower) ||
      entry.transliterationLower.includes(lower)
    if (!hit) continue
    out.push(toHitWithDisplayTranslation(entry, index, i, trimmed))
    if (out.length >= maxHits) break
  }
  return out
}

/**
 * SI ayah text when it can show the match; otherwise the same-ayah word-gloss
 * line when that can. Falls back to SI when neither hosts a highlight.
 */
function toHitWithDisplayTranslation(
  entry: WordSearchIndexEntry,
  index: WordSearchIndexEntry[],
  at: number,
  query: string,
): WordSearchHit {
  const base = toHit(entry)
  if (highlightNeedle(entry.ayahTranslation, query, entry.translation) != null) {
    return base
  }
  const glossLine = sameAyahGlossLine(index, at)
  if (highlightNeedle(glossLine, query, entry.translation) != null) {
    return { ...base, ayahTranslation: glossLine }
  }
  return base
}

/** Space-joined English glosses for every word of the same ayah as [at]. */
export function sameAyahGlossLine(
  index: WordSearchIndexEntry[],
  at: number,
): string {
  if (at < 0 || at >= index.length) return ''
  const anchor = index[at]!
  let lo = at
  while (
    lo > 0 &&
    index[lo - 1]!.surahId === anchor.surahId &&
    index[lo - 1]!.ayahNumber === anchor.ayahNumber
  ) {
    lo--
  }
  let hi = at
  while (
    hi + 1 < index.length &&
    index[hi + 1]!.surahId === anchor.surahId &&
    index[hi + 1]!.ayahNumber === anchor.ayahNumber
  ) {
    hi++
  }
  const parts: string[] = []
  for (let i = lo; i <= hi; i++) parts.push(index[i]!.translation)
  return parts.join(' ')
}

/** How many index rows to scan before yielding to the event loop. */
export const WORD_SEARCH_CHUNK = 4_000

/**
 * Cooperative match — same results as [matchWordSearch], but yields every
 * [WORD_SEARCH_CHUNK] rows so typing stays responsive on the main thread.
 * Callers pass [isCancelled] to drop stale queries (Android `collectLatest`).
 */
export async function matchWordSearchAsync(
  index: WordSearchIndexEntry[],
  query: string,
  maxHits = WORD_SEARCH_MAX_HITS,
  isCancelled: () => boolean = () => false,
): Promise<WordSearchHit[]> {
  const trimmed = query.trim()
  if (!isWordSearchQuery(trimmed)) return []
  const arabicNorm = normalizeArabicForSearch(trimmed)
  const lower = trimmed.toLowerCase()
  const wantArabic = arabicNorm.length >= WORD_SEARCH_MIN_QUERY_LENGTH
  const out: WordSearchHit[] = []
  for (let i = 0; i < index.length; i++) {
    if (i > 0 && i % WORD_SEARCH_CHUNK === 0) {
      if (isCancelled()) return out
      await yieldToEventLoop()
      if (isCancelled()) return out
    }
    const entry = index[i]!
    const hit =
      (wantArabic && entry.arabicNorm.includes(arabicNorm)) ||
      entry.translationLower.includes(lower) ||
      entry.transliterationLower.includes(lower)
    if (!hit) continue
    out.push(toHitWithDisplayTranslation(entry, index, i, trimmed))
    if (out.length >= maxHits) break
  }
  return out
}

function yieldToEventLoop(): Promise<void> {
  return new Promise((resolve) => {
    if (typeof requestAnimationFrame === 'function') {
      requestAnimationFrame(() => resolve())
    } else {
      setTimeout(resolve, 0)
    }
  })
}

export function sectionWordSearchHits(
  hits: WordSearchHit[],
  expandedSurahIds: Set<number>,
  previewLimit = WORD_SEARCH_PREVIEW_LIMIT,
): SurahWordSearchSection[] {
  if (hits.length === 0) return []
  const grouped = new Map<number, WordSearchHit[]>()
  for (const hit of hits) {
    const list = grouped.get(hit.surahId)
    if (list) list.push(hit)
    else grouped.set(hit.surahId, [hit])
  }
  const sections: SurahWordSearchSection[] = []
  for (const [surahId, surahHits] of grouped) {
    const expanded = expandedSurahIds.has(surahId)
    const visible =
      expanded || surahHits.length <= previewLimit
        ? surahHits
        : surahHits.slice(0, previewLimit)
    const first = surahHits[0]!
    sections.push({
      surahId,
      surahNameTransliteration: first.surahNameTransliteration,
      surahNameArabic: first.surahNameArabic,
      hits: visible,
      totalCount: surahHits.length,
      expanded,
      hiddenCount: Math.max(0, surahHits.length - visible.length),
    })
  }
  return sections
}

export interface AyahTextSpan {
  text: string
  highlighted: boolean
}

export function ayahHighlightSpans(
  ayahText: string,
  position: number,
  fallbackWord: string,
): AyahTextSpan[] {
  if (!ayahText) return []
  const tokens = ayahText.split(/\s+/).filter((t) => t.length > 0)
  if (position >= 1 && position <= tokens.length) {
    const spans: AyahTextSpan[] = []
    tokens.forEach((token, index) => {
      if (index > 0) spans.push({ text: ' ', highlighted: false })
      spans.push({ text: token, highlighted: index + 1 === position })
    })
    return spans
  }
  if (!fallbackWord) return [{ text: ayahText, highlighted: false }]
  const spans: AyahTextSpan[] = []
  let start = 0
  let i = ayahText.indexOf(fallbackWord)
  if (i < 0) return [{ text: ayahText, highlighted: false }]
  while (i >= 0) {
    if (i > start) {
      spans.push({ text: ayahText.slice(start, i), highlighted: false })
    }
    spans.push({ text: fallbackWord, highlighted: true })
    start = i + fallbackWord.length
    i = ayahText.indexOf(fallbackWord, start)
  }
  if (start < ayahText.length) {
    spans.push({ text: ayahText.slice(start), highlighted: false })
  }
  return spans
}

/** How many whole words of context to keep on each side of a search hit. */
export const SNIPPET_WORDS_BEFORE = 8
export const SNIPPET_WORDS_AFTER = 14

/**
 * Builds spans for an English search snippet: a short window centered on the
 * match (query or word gloss) with the match highlighted.
 */
export function englishTranslationHighlightSpans(
  ayahTranslation: string,
  query: string,
  wordGloss: string,
): AyahTextSpan[] {
  if (!ayahTranslation) return []
  const needle = highlightNeedle(
    ayahTranslation,
    query.trim(),
    wordGloss.trim(),
  )
  const snippet = windowAroundMatch(
    ayahTranslation,
    needle,
    SNIPPET_WORDS_BEFORE,
    SNIPPET_WORDS_AFTER,
  )
  if (!needle) return [{ text: snippet, highlighted: false }]
  return highlightAllOccurrences(snippet, needle)
}

/**
 * Trims [text] to roughly [wordsBefore]…[wordsAfter] words around the first
 * occurrence of [needle], adding an ellipsis when the ends were cut.
 */
export function windowAroundMatch(
  text: string,
  needle: string | null,
  wordsBefore = SNIPPET_WORDS_BEFORE,
  wordsAfter = SNIPPET_WORDS_AFTER,
): string {
  if (!needle || !text) return text
  const matchStart = text.toLowerCase().indexOf(needle.toLowerCase())
  if (matchStart < 0) return text
  const matchEnd = matchStart + needle.length
  const words = [...text.matchAll(/\S+/g)]
  if (words.length === 0) return text
  let matchWord = 0
  for (let i = 0; i < words.length; i++) {
    const m = words[i]!
    const start = m.index ?? 0
    const end = start + m[0]!.length
    if (matchStart < end && matchEnd > start) {
      matchWord = i
      break
    }
  }
  const from = Math.max(0, matchWord - wordsBefore)
  const to = Math.min(words.length - 1, matchWord + wordsAfter)
  const startChar = words[from]!.index ?? 0
  const endWord = words[to]!
  const endChar = (endWord.index ?? 0) + endWord[0]!.length
  const core = text.slice(startChar, endChar).trim()
  const prefix = from > 0 ? '…' : ''
  const suffix = to < words.length - 1 ? '…' : ''
  return prefix + core + suffix
}

/** Prefers the typed query when present; otherwise the word gloss / a token. */
export function highlightNeedle(
  haystack: string,
  query: string,
  wordGloss: string,
): string | null {
  if (query && haystack.toLowerCase().includes(query.toLowerCase())) {
    return query
  }
  if (wordGloss && haystack.toLowerCase().includes(wordGloss.toLowerCase())) {
    return wordGloss
  }
  const tokens = wordGloss
    .split(/[\s,;:]+/)
    .map((t) => t.trim().replace(/^[([{"']+|[)\]}"']+$/g, ''))
    .filter((t) => t.length >= 3)
    .sort((a, b) => b.length - a.length)
  for (const token of tokens) {
    if (haystack.toLowerCase().includes(token.toLowerCase())) return token
  }
  return null
}

function highlightAllOccurrences(text: string, needle: string): AyahTextSpan[] {
  const spans: AyahTextSpan[] = []
  const lowerText = text.toLowerCase()
  const lowerNeedle = needle.toLowerCase()
  let start = 0
  let i = lowerText.indexOf(lowerNeedle)
  if (i < 0) return [{ text, highlighted: false }]
  while (i >= 0) {
    if (i > start) spans.push({ text: text.slice(start, i), highlighted: false })
    const end = i + needle.length
    spans.push({ text: text.slice(i, end), highlighted: true })
    start = end
    i = lowerText.indexOf(lowerNeedle, start)
  }
  if (start < text.length) spans.push({ text: text.slice(start), highlighted: false })
  return spans
}

export interface SurahFilterResult {
  surahs: { id: number; nameArabic: string; nameTransliteration: string; nameTranslation: string; ayahCount: number }[]
  ayahTarget: number | null
}

/** Home surah filter — mirrors Android `filterSurahs`. */
export function filterSurahs<T extends {
  id: number
  nameArabic: string
  nameTransliteration: string
  nameTranslation: string
  ayahCount: number
}>(surahs: T[], query: string): { surahs: T[]; ayahTarget: number | null } {
  const reference = parseAyahReference(query)
  if (!query.trim()) return { surahs, ayahTarget: null }
  if (reference != null) {
    const surah = surahs.find((s) => s.id === reference.surah)
    const ayahInRange =
      reference.ayah == null ||
      (surah != null && reference.ayah >= 1 && reference.ayah <= surah.ayahCount)
    if (surah != null && ayahInRange) {
      return { surahs: [surah], ayahTarget: reference.ayah }
    }
    return { surahs: [], ayahTarget: null }
  }
  const trimmed = query.trim()
  const lower = trimmed.toLowerCase()
  return {
    surahs: surahs.filter(
      (s) =>
        s.nameTransliteration.toLowerCase().includes(lower) ||
        s.nameTranslation.toLowerCase().includes(lower) ||
        s.nameArabic.includes(trimmed) ||
        String(s.id) === trimmed,
    ),
    ayahTarget: null,
  }
}
